src/app-logic/url-handling.js

```diff
@@ -180,68 +203,86 @@ export function urlStateToUrlObject(urlState: UrlState): UrlObject {
+
+    if (urlState.profileSpecific.full.timelineType === 'stack') {
+      // The default is the category view, so only add it to the URL if it's the
+      // stack view.
+      baseQuery.timelineType = 'stack';
+    }
+
+    let localTrackOrderByPid = '';
+    for (const [pid, trackOrder] of urlState.profileSpecific.full
+      .localTrackOrderByPid) {
+      if (trackOrder.length > 0) {
+        localTrackOrderByPid += `${String(pid)}-` + trackOrder.join('-') + '~';
+      }
+    }
+    baseQuery.localTrackOrderByPid = localTrackOrderByPid || undefined;
+  } else {
+    baseQuery = ({}: ActiveTabProfileSpecificBaseQueryShape);
+  }
+
+  baseQuery = ({
+    ...baseQuery,
     range:
       stringifyCommittedRanges(urlState.profileSpecific.committedRanges) ||
       undefined,
     thread: selectedThread === null ? undefined : selectedThread.toString(),
-    globalTrackOrder:
-      urlState.profileSpecific.globalTrackOrder.join('-') || undefined,
     file: urlState.pathInZipFile || undefined,
     profiles: urlState.profilesToCompare || undefined,
     v: CURRENT_URL_VERSION,
     profileName: urlState.profileName || undefined,
     showTabOnly1: urlState.showTabOnly || undefined,
-  };
-
-  // Add the parameter hiddenGlobalTracks only when needed.
-  if (urlState.profileSpecific.hiddenGlobalTracks.size > 0) {
-    query.hiddenGlobalTracks = [
-      ...urlState.profileSpecific.hiddenGlobalTracks,
-    ].join('-');
-  }
-
-  let hiddenLocalTracksByPid = '';
-  for (const [pid, tracks] of urlState.profileSpecific.hiddenLocalTracksByPid) {
-    if (tracks.size > 0) {
-      hiddenLocalTracksByPid += [pid, ...tracks].join('-') + '~';
-    }
-  }
-  if (hiddenLocalTracksByPid.length > 0) {
-    // Only add to the query string if something was actually hidden.
-    // Also, slice off the last '~'.
-    query.hiddenLocalTracksByPid = hiddenLocalTracksByPid.slice(0, -1);
-  }
+  }: BaseQueryShape);
 
-  if (urlState.profileSpecific.timelineType === 'stack') {
-    // The default is the category view, so only add it to the URL if it's the
-    // stack view.
-    query.timelineType = 'stack';
-  }
-
-  let localTrackOrderByPid = '';
-  for (const [pid, trackOrder] of urlState.profileSpecific
-    .localTrackOrderByPid) {
-    if (trackOrder.length > 0) {
-      localTrackOrderByPid += `${String(pid)}-` + trackOrder.join('-') + '~';
-    }
-  }
-  query.localTrackOrderByPid = localTrackOrderByPid || undefined;
-
-  // Depending on which tab is active, also show tab-specific query parameters.
+  // Depending on which panel is active, also show tab-specific query parameters.
+  let query: QueryShape;
   const selectedTab = urlState.selectedTab;
   switch (selectedTab) {
     case 'stack-chart':
     case 'flame-graph':
     case 'calltree': {
+      if (selectedTab === 'stack-chart') {
+        // Stack chart uses all of the CallTree's query strings but also has an
+        // additional query string.
+        query = (baseQuery: StackChartQueryShape);
```

Comment:
Doing the casting for individual branches here (and in other branches), to make the flow work after union change.